rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -176,11 +195,27 @@ class PublisherBase
     const rcl_publisher_options_t & intra_process_options);
 
 protected:
+  template<typename EventCallbackT>
+  void
+  add_event_handler(
+    const EventCallbackT & callback,
+    const rcl_publisher_event_type_t event_type)
+  {
+    auto handler = std::make_shared<QOSEventHandler<EventCallbackT>>(
+      callback,
+      rcl_publisher_event_init,
+      &publisher_handle_,
+      event_type);
+    event_handlers_.emplace_back(handler);
+  }
+
   std::shared_ptr<rcl_node_t> rcl_node_handle_;
 
   rcl_publisher_t publisher_handle_ = rcl_get_zero_initialized_publisher();
   rcl_publisher_t intra_process_publisher_handle_ = rcl_get_zero_initialized_publisher();
 
+  std::vector<std::shared_ptr<QOSEventHandlerBase>> event_handlers_;
```

Comment:
Again, use `rclcpp::`.

Reply:
Do you intend to have more than one callback per type? If not, maybe use an unordered map? Also, has any thought been given to the idea of removing callbacks after construction, without destroying the publisher, and/or adding callbacks after construction, something like:

```c++
auto pub = node->create_publisher<...>(...);
pub->on_deadline_missed([](...){...});
```

Reply:
Will use the qualified name `rclcpp::QOSEventHandlerBase`.

There were thought given to adding/removing callbacks after construction, but the current design doc ultimately chose to omit that use case. There will also not be more than one callback per type.

That said, I don't think using an `unordered_map` will bring any advantages, and it is a more complex or heavier type than a `vector`. If in the future we decide to add the ability to add/remove callbacks after construction, then `unordered_map` will be worth considering.

Reply:
Leaving it for a future improvement is fine by me.

Yeah, you'd only want the map if you needed to remove them later. Identifying them by index in the vector or by pointer address wouldn't work as well as an explicit integer key in a map.